File: admin/src/api/media.ts
import { apiRequest } from "#/api/client.ts";

export type MediaKind = "user" | "category" | "product";

export type MediaAsset = {
  pid: string;
  publicId: string;
  secureUrl: string;
  folder: string;
  status: string;
  format: string | null;
  bytes: number | null;
  width: number | null;
  height: number | null;
};

type UploadSignature = {
  assetPid: string;
  cloudName: string;
  apiKey: string;
  timestamp: number;
  folder: string;
  publicId: string;
  signature: string;
};

type CloudinaryUploadResponse = {
  asset_id?: string;
  bytes?: number;
  format?: string;
  height?: number;
  public_id?: string;
  secure_url?: string;
  width?: number;
};

export type UploadedMedia = {
  assetPid: string;
  imageLink: string;
};

export function listMediaAssets() {
  return apiRequest<MediaAsset[]>("/media", {
    fallback: "Unable to load media assets",
  });
}

export function signMediaUpload(kind: MediaKind, checksum?: string) {
  return apiRequest<UploadSignature>("/media/sign", {
    method: "POST",
    body: JSON.stringify({ kind, checksum }),
    fallback: "Unable to prepare media upload",
  });
}

export async function uploadMedia(file: File, kind: MediaKind): Promise<UploadedMedia> {
  const signature = await signMediaUpload(kind);
  const body = new FormData();
  body.set("file", file);
  body.set("api_key", signature.apiKey);
  body.set("timestamp", String(signature.timestamp));
  body.set("folder", signature.folder);
  body.set("public_id", signature.publicId);
  body.set("signature", signature.signature);

  const response = await fetch(
    `https://api.cloudinary.com/v1_1/${signature.cloudName}/image/upload`,
    { method: "POST", body },
  );

  if (!response.ok) {
    throw new Error(`Unable to upload ${kind} image`);
  }

  const result = (await response.json()) as CloudinaryUploadResponse;
  const expectedPublicIds = new Set([
    signature.publicId,
    `${signature.folder}/${signature.publicId}`,
  ]);
  if (!result.secure_url || !result.public_id || !expectedPublicIds.has(result.public_id)) {
    throw new Error("Cloudinary returned an invalid media asset");
  }

  const asset = await apiRequest<MediaAsset>(`/media/${signature.assetPid}/finalize`, {
    method: "PUT",
    body: JSON.stringify({
      publicId: result.public_id,
      assetId: result.asset_id,
      secureUrl: result.secure_url,
      format: result.format,
      bytes: result.bytes,
      width: result.width,
      height: result.height,
    }),
    fallback: "Unable to register uploaded media",
  });

  return { assetPid: asset.pid, imageLink: asset.secureUrl };
}

export function uploadAvatarImage(file: File) {
  return uploadMedia(file, "user");
}

export function uploadCategoryImage(file: File) {
  return uploadMedia(file, "category");
}

export function uploadProductImage(file: File) {
  return uploadMedia(file, "product");
}

File: service/tests/requests/media.rs
use std::{env, ffi::OsString};

use axum::http::{HeaderValue, StatusCode};
use axum_test::TestServer;
use serde_json::{Value, json};
use serial_test::serial;
use sha1::{Digest, Sha1};

use crate::utils;

const TEST_CLOUD_NAME: &str = "test-cloud";
const TEST_API_KEY: &str = "test-api-key";
const TEST_API_SECRET: &str = "test-api-secret";

struct CloudinaryEnvGuard {
    previous: Vec<(&'static str, Option<OsString>)>,
}

impl CloudinaryEnvGuard {
    fn set() -> Self {
        let variables = [
            ("APP_CLOUDINARY_CLOUD_NAME", TEST_CLOUD_NAME),
            ("APP_CLOUDINARY_API_KEY", TEST_API_KEY),
            ("APP_CLOUDINARY_API_SECRET", TEST_API_SECRET),
        ];
        let previous = variables
            .iter()
            .map(|(key, _)| (*key, env::var_os(key)))
            .collect();

        for (key, value) in variables {
            // SAFETY: Media request tests are serial and restore variables on drop.
            unsafe { env::set_var(key, value) };
        }

        Self { previous }
    }
}

impl Drop for CloudinaryEnvGuard {
    fn drop(&mut self) {
        for (key, value) in &self.previous {
            // SAFETY: Restore the process environment modified by this serial test.
            unsafe {
                match value {
                    Some(value) => env::set_var(key, value),
                    None => env::remove_var(key),
                }
            }
        }
    }
}

async fn access_token(server: &TestServer) -> HeaderValue {
    utils::login_users(
        server,
        &json!({"email": "admin@silk.com", "password": "Password"}),
    )
    .await
    .access_token
}

fn auth_header(token: HeaderValue) -> (axum::http::HeaderName, HeaderValue) {
    utils::auth_header(token)
}

fn expected_signature(payload: &Value) -> String {
    let canonical = format!(
        "folder={}&public_id={}&timestamp={}",
        payload["folder"].as_str().unwrap(),
        payload["publicId"].as_str().unwrap(),
        payload["timestamp"].as_i64().unwrap()
    );
    let mut hasher = Sha1::new();
    hasher.update(canonical.as_bytes());
    hasher.update(TEST_API_SECRET.as_bytes());
    hex::encode(hasher.finalize())
}

#[tokio::test]
#[serial]
async fn sign_finalize_list_and_reconcile_media_assets() {
    let _cloudinary = CloudinaryEnvGuard::set();

    crate::request(|server, ctx| async move {
        crate::seed_data(ctx.db())
            .await
            .expect("Failed to seed data");

        let token = access_token(&server).await;
        let (header, value) = auth_header(token.clone());
        let sign_response = server
            .post("/media/sign")
            .add_header(header, value)
            .json(&json!({"kind": "product", "checksum": "test-checksum"}))
            .await;

        assert_eq!(sign_response.status_code(), StatusCode::OK);
        let signed: Value = sign_response.json();
        assert_eq!(signed["cloudName"], json!(TEST_CLOUD_NAME));
        assert_eq!(signed["apiKey"], json!(TEST_API_KEY));
        assert_eq!(signed["folder"], json!("silk/products"));
        assert!(uuid::Uuid::parse_str(signed["publicId"].as_str().unwrap()).is_ok());
        assert_eq!(signed["signature"], json!(expected_signature(&signed)));

        let asset_pid = signed["assetPid"].as_str().unwrap();
        let public_id = format!(
            "{}/{}",
            signed["folder"].as_str().unwrap(),
            signed["publicId"].as_str().unwrap()
        );
        let secure_url =
            format!("https://res.cloudinary.com/{TEST_CLOUD_NAME}/image/upload/{public_id}.png");

        let (header, value) = auth_header(token.clone());
        let invalid_finalize = server
            .put(&format!("/media/{asset_pid}/finalize"))
            .add_header(header, value)
            .json(&json!({
                "publicId": public_id,
                "secureUrl": "https://cdn.example.com/not-cloudinary.png"
            }))
            .await;
        assert_eq!(invalid_finalize.status_code(), StatusCode::BAD_REQUEST);

        let (header, value) = auth_header(token.clone());
        let finalize_response = server
            .put(&format!("/media/{asset_pid}/finalize"))
            .add_header(header, value)
            .json(&json!({
                "publicId": public_id,
                "assetId": "cloudinary-asset-id",
                "secureUrl": secure_url,
                "format": "png",
                "bytes": 1024,
                "width": 640,
                "height": 480,
                "checksum": "test-checksum"
            }))
            .await;
        assert_eq!(finalize_response.status_code(), StatusCode::OK);
        let finalized: Value = finalize_response.json();
        assert_eq!(finalized["status"], json!("active"));
        assert_eq!(finalized["publicId"], json!(public_id));

        let (header, value) = auth_header(token.clone());
        let list_response = server.get("/media").add_header(header, value).await;
        assert_eq!(list_response.status_code(), StatusCode::OK);
        let assets: Value = list_response.json();
        assert!(
            assets
                .as_array()
                .unwrap()
                .iter()
                .any(|asset| asset["pid"] == json!(asset_pid))
        );

        let (header, value) = auth_header(token);
        let reconcile_response = server
            .post("/media/reconcile")
            .add_header(header, value)
            .await;
        assert_eq!(reconcile_response.status_code(), StatusCode::OK);
        let quarantined: Value = reconcile_response.json();
        assert!(
            quarantined
                .as_array()
                .unwrap()
                .iter()
                .any(|asset| asset["pid"] == json!(asset_pid) && asset["status"] == "quarantined")
        );
    })
    .await;
}

#[tokio::test]
#[serial]
async fn media_endpoints_require_authentication() {
    crate::request(|server, ctx| async move {
        crate::seed_data(ctx.db())
            .await
            .expect("Failed to seed data");

        let response = server
            .post("/media/sign")
            .json(&json!({"kind": "product"}))
            .await;

        assert_eq!(response.status_code(), StatusCode::UNAUTHORIZED);
    })
    .await;
}
